core/domain/exp_jobs_one_off.py

```diff
@@ -490,3 +493,109 @@ def reduce(key, values):
+                'Exploration %s failed non-strict validation: %s' %
+                (item.id, e))
+            return
+        if (item.states_schema_version >=
+                AUDIO_DURATION_SECS_MIN_STATE_SCHEMA_VERSION
+                and item.states_schema_version <=
+                feconf.CURRENT_STATE_SCHEMA_VERSION):
+            # Go through each exploration state to find voiceover recordings.
+            # For batching them as commits per explorations.
+            commit_cmds = []
+            change_count = 0
+            unchanged_count = 0
+            failed_count = 0
+            for state, state_value in item.states.items():
+                voiceovers_mapping = (state_value['recorded_voiceovers']
+                                      ['voiceovers_mapping'])
+                language_codes_to_audio_metadata = voiceovers_mapping.values()
+                for language_codes in language_codes_to_audio_metadata:
+                    for audio_metadata in language_codes.values():
+                        # Get files using the filename.
+                        filename = audio_metadata['filename']
+                        if audio_metadata['duration_secs'] == 0.0:
+                            try:
+                                fs = (fs_domain.AbstractFileSystem(
+                                    fs_domain.GcsFileSystem(
+                                        AUDIO_ENTITY_TYPE,
+                                        item.id)))
+                                raw = (
+                                    fs.get('%s/%s' % (AUDIO_FILE_PREFIX,
+                                                      filename)))
+                                # Get the audio-duration from file use Mutagen.
+                                tempbuffer = python_utils.string_io()
+                                tempbuffer.write(raw)
+                                tempbuffer.seek(0)
+                                # Loads audio metadata with Mutagen.
+                                audio = mp3.MP3(tempbuffer)
+                                tempbuffer.close()
+                                # Fetch the audio file duration from the Mutagen
+                                # metadata.
+                                audio_metadata['duration_secs'] = (
+                                    audio.info.length)
+                                change_count += 1
+                            except Exception as e:
+                                logging.error(
+                                    'MP3 audio file exception for file %s ,'
+                                    'EXP_ID: %s,'
+                                    'STATE_NAME: %s,'
+                                    'REASON: %s' %
+                                    (filename, item.id, state, e))
+                                failed_count += 1
+                        else:
+                            unchanged_count += 1
+                if change_count > 0:
```

Comment:
This seems wrong -- the change_count here should be state-specific, right?

Reply:
Yeah that's how it's supposed to be. Maybe I should make this more clear.

Reply:
clear this up done.